server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/event/ComputeMessagePreviewListener.java

```diff
@@ -0,0 +1,101 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.event;
+
+import java.util.List;
+
+import javax.inject.Inject;
+
+import org.apache.james.jmap.api.preview.MessagePreviewStore;
+import org.apache.james.jmap.api.preview.Preview;
+import org.apache.james.jmap.draft.model.message.view.MessageFullViewFactory;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageIdManager;
+import org.apache.james.mailbox.events.Event;
+import org.apache.james.mailbox.events.Group;
+import org.apache.james.mailbox.events.MailboxListener;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.FetchGroupImpl;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MessageResult;
+import org.parboiled.common.ImmutableList;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+import reactor.core.publisher.Mono;
+
+public class ComputeMessagePreviewListener implements MailboxListener.GroupMailboxListener {
+    public static class ComputeMessagePreviewListenerGroup extends Group {
+
+    }
+
+    private static final Group GROUP = new ComputeMessagePreviewListenerGroup();
+
+    private final MessageIdManager messageIdManager;
+    private final MessagePreviewStore messagePreviewStore;
+    private final MessageFullViewFactory messageFullViewFactory;
+    private final MailboxManager mailboxManager;
+
+    @Inject
+    public ComputeMessagePreviewListener(MailboxManager mailboxManager, MessageIdManager messageIdManager,
+                                            MessagePreviewStore messagePreviewStore, MessageFullViewFactory messageFullViewFactory) {
+        this.mailboxManager = mailboxManager;
+        this.messageIdManager = messageIdManager;
+        this.messagePreviewStore = messagePreviewStore;
+        this.messageFullViewFactory = messageFullViewFactory;
+    }
+
+    @Override
+    public Group getDefaultGroup() {
+        return GROUP;
+    }
+
+    @Override
+    public void event(Event event) throws MailboxException {
+        if (event instanceof Added) {
+            MailboxSession session = mailboxManager.createSystemSession(event.getUsername());
+            handleAddedEvent((Added) event, session);
+        }
+    }
+
+    private void handleAddedEvent(Added addedEvent, MailboxSession session) throws MailboxException {
+        List<MessageId> messageIds = addedEvent.getUids()
+            .stream()
+            .map(uid -> addedEvent.getMetaData(uid).getMessageId())
+            .distinct()
+            .collect(Guavate.toImmutableList());
```

Comment:
BTW, I didn't see tests about this `distinct` thing

Reply:
Why a Set? We need a List of MessageId, not a Set to call `getMessages` from `MessageIdManager` ?

Reply:
Do you mean I should put the function (extracted to `getEventMessageIds` now in last commits) in package private `@VisibleForTesting` to test that behavior?

Reply:
what I say is: it looks like (but of course I may be wrong) removing the `.distinct` will not make any test fail.
If it's the case, It means you lack (at least) at test

Reply:
This would happen if a message end up being sored several time in the same mailbox (duplicated uid for a single messageId)

Let's add a unit test for this.

Reply:
In the end added event API is just not good enough.